### PR TITLE
[jit] fix dce over loops

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22632 [jit] fix dce over loops**

We need re-run the marking pass over loop sub-blocks until they converge
to a fixed point. Thanks to @chillee for catching this bug!

Differential Revision: [D16184469](https://our.internmc.facebook.com/intern/diff/D16184469)

### CODE REVIEW

File: torch/csrc/jit/passes/dead_code_elimination.cpp [hunk 1]
```diff
@@ -147,15 +183,23 @@ class DeadCodeEliminator {
 
     for (auto it = block->nodes().rbegin(); it != block->nodes().rend(); ++it) {
       auto node = *it;
-      for (auto subBlock : node->blocks()) {
-        mark(subBlock);
+      if (node->kind() == prim::Loop) {
```
Comment (Chillee) [inline torch/csrc/jit/passes/dead_code_elimination.cpp]: Do we need to check for `c10::onnx::loop` here like we do in the other place?

**Review by Chillee — APPROVED** [review]
Otherwise looks pretty straightforward and correct to me.

File: torch/csrc/jit/passes/dead_code_elimination.cpp [hunk 2]
```diff
@@ -100,7 +100,8 @@ class DeadCodeEliminator {
       // If there's no outer node, we're looking at the graph's top-level
       // return block. We consider all graph outputs to be "used", so just mark
       // this node normally.
-      return mark(node);
+      mark(node);
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/dead_code_elimination.cpp]: I don't quite like the fact that `markReturnNode` isn't changed to `bool`. I get that we only call `mark` for the top-level (function return) and in this case we aren't in a loop, so it doesn't matter, and if we are in a loop we make live only body outputs. When we'll get early returns we might revisiting this and make `markReturnNode` to actually return a `bool`

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/passes/dead_code_elimination.cpp (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/passes/dead_code_elimination.cpp]: onnx is purely functional and has no aliasing, so we wouldn't need the special behavior

**Review by zdevito — APPROVED** [review]
Seems fine.

**Pull request closed without merging**